File: src/pages/posts.rs
use tmpl::{Tmpl, apply_field};

use crate::utils::Post;

pub struct RenderedPost {
    pub file_path: String,
    pub post: Template,
}

pub struct Template {
    title: String,
    pub_date: String,
    children: String,
}
impl Tmpl for Template {
    fn render(&self, template: &str) -> String {
        apply_field!(self, title, template);
        apply_field!(self, pub_date, template);
        apply_field!(self, children, template);
        template
    }
}

pub fn posts(posts: &[Post]) -> Vec<RenderedPost> {
    posts
        .iter()
        .map(|post| {
            let post = post.clone();
            let file_path = format!("{}.html", post.slug);

            let post = Template {
                title: post.title,
                pub_date: post.date.strftime("%F").to_string(),
                children: post.content,
            };

            RenderedPost { file_path, post }
        })
        .collect()
}

File: src/pages/posts.go
package pages

import (
	"bytes"
	"fmt"
	"text/template"
	"time"

	"tangled.org/zerolimits.dev/znak"
	utils "tangled.sh/zerolimits.dev/blog/src"
)

type renderedPost struct {
	FilePath string
	Post     string
}

func Posts() ([]renderedPost, error) {
	type tmplVals struct {
		Title    string
		PubDate  string
		Children string
	}

	posts, err := utils.GetPosts()
	if err != nil {
		return nil, err
	}
	tmpl, err := template.ParseFiles("src/pages/posts.html")
	if err != nil {
		return nil, err
	}

	renderedPosts := []renderedPost{}
	for _, post := range posts {
		filePath := fmt.Sprintf("%s.html", post.Slug)

		theme, err := utils.GetTheme()
		if err != nil {
			return nil, err
		}
		content, err := znak.Render(post.Content, theme)
		if err != nil {
			return nil, err
		}

		postPage := bytes.Buffer{}
		vals := tmplVals{
			Title:    post.Title,
			PubDate:  post.Date.Format(time.DateOnly),
			Children: content,
		}

		if err = tmpl.Execute(&postPage, vals); err != nil {
			return nil, err
		}
		renderedPosts = append(renderedPosts, renderedPost{
			FilePath: filePath,
			Post:     postPage.String(),
		})
	}

	return renderedPosts, nil
}
